File: Linked_list.c
#include<stdlib.h>
#include<stdbool.h>
#include<stdio.h>
#include"Linked_List.h"

struct Node *head = NULL;

void insert_first(int value){
    struct Node *link = (struct Node*) malloc(sizeof(struct Node));
    link -> value = value;
    link -> next = head;
    head = link;
}

struct Node* get_last(){
    struct Node* var = head;
    for(var; var -> next; var = var -> next){}
    return var;
}

int list_size(){
    int size = 0;
    for(struct Node *var = head; var != NULL; size++, var = var -> next){}
    return size;
}

void insert_last(int value){
    struct Node *link = (struct Node*) malloc(sizeof(struct Node));
    link -> value = value;
    link -> next = NULL;
    if(head == NULL){
        head = link;
    }
    else{
        struct Node *last = get_last();
        last -> next = link;
    }
}

void print_list(){
    printf("[ ");
    struct Node *var = head;
    printf("%d", var -> value);
    for(var = var -> next; var != NULL; var = var -> next){
        printf(", %d", var -> value);
    }
    printf(" ]\n");
}

struct Node* delete(int value){
    for(struct Node *var = head, *previous = NULL; var != NULL; previous = var, var = var -> next){
        if(var -> value == value){
            if(var == head){
                head = head -> next;
            }
            else{
                previous -> next = var -> next;
                return var;
            }
        }
    }
    return NULL;
}

int insert(int data, int position){
    if(list_size() >= position){
        struct Node *var1 = (struct Node*) malloc(sizeof(struct Node));
        var1 -> value = data;
        if(0 == position){
            var1 -> next = head;
            head = var1;
        }
        else if(list_size() == position){
            insert_last(data);
        }
        else{
            int count = 0;
            for(struct Node* var = head, *previous = NULL; var != NULL; previous = var, var = var -> next, count++){
                if(count == position){
                    previous -> next = var1;
                    var1 -> next = var;
                }
            }
        }
        return 0;
    }
    else{
        return 1;
    }
}

void insert_after(int data1, int data2){
    for(struct Node *var = head; var != NULL; var = var -> next){
        if(var -> value == data1){
            struct Node *node = (struct Node*) malloc(sizeof(struct Node));
            node -> value = data2;
            node -> next = var -> next;
            var -> next = node;
            break;
        }
    }
}

void bubble_sort(){
    for(int i = 0; i < list_size() - 1; i++){
        for(struct Node *var = head; var -> next; var = var -> next){
            if(var -> value > var -> next -> value){
                int temp = var -> value;
                var -> value = var -> next -> value;
                var -> next -> value = temp;
            }
        }
    }
}

void replace(int data, int new_data){
    for(struct Node* var = head, *previous = NULL; var != NULL; previous = var, var = var -> next){
        if(var -> value == data){
            struct Node *node = (struct Node*) malloc(sizeof(struct Node));
            node -> value = new_data;
            node -> next = var -> next;
            previous -> next = node;
        }
    }
}

File: data structure/Linked_List.java
class Node{
    int value;
    Node next;
    public Node(int value){
        this.value = value;
        this.next = null;
    }
    public boolean has_next(){
        return this.next != null;
    }
    public Node get_next(){
        return this.next;
    }
    public int get_value(){
        return this.value;
    }
    public void set_next(Node node){
        this.next = node;
    }
}

public class Linked_List{
    Node node_head = null;
    public Linked_List(){}
    public void add_last(int value){
        Node new_node = new Node(value);
        if(node_head == null){
            node_head = new Node(value);
        }
        else{
            Node var;
            for(var = node_head; var.has_next(); var = var.get_next()){}
            var.set_next(new_node);
        }
    }
    public void traverse(){
        System.out.print("[");
        for(Node var = node_head; var != null; var = var.get_next()){
            System.out.print(" " + var.get_value());
        }
        System.out.println(" ]");
    }
    public void remove(int value){
        if(node_head.get_value() == value){
            node_head = node_head.get_next();
        }
        else{
            for(Node current = node_head, previous = null; current != null; previous = current, current = current.get_next()){
                if(current.get_value() == value){
                    previous.set_next(current.get_next());
                }
            }
        }
    }
    public void insert(int data, int position){
        int count = 0;
        Node new_data = new Node(data);
        if(position == 0){
            new_data.set_next(node_head);
            node_head = new_data;
        }
        else{
            for(Node var = node_head, previous = null; var != null; previous = var, var = var.get_next(), count++){
                if(position == count){
                    new_data.set_next(var);
                    previous.set_next(new_data);
                    break;
                }
            }
        }
    }
    public void insert_after(int data1, int data2){
        Node new_data = new Node(data2);
        if(node_head.get_value() == data1){
            new_data.set_next(node_head.get_next());
            node_head.set_next(new_data);
        }
        else{
            for(Node var = node_head; var != null; var = var.get_next()){
                if(var.get_value() == data1){
                    new_data.set_next(var.get_next());
                    var.set_next(new_data);
                    break;
                }
            }
        }
    }
}